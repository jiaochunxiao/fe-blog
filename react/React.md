## React

### JSX

#### JSX 防止注入攻击

React DOM在渲染所有输入内容之前，默认会进行转义。它可以确保在你的应用中，永远不会注入那些并非自己明确编写的内容。所有的内容在渲染之前都被转换成了字符串。这样可以有效地防止 XSS（cross-site-scripting, 跨站脚本）攻击。

### 元素渲染

元素是构成 React 应用的最小砖块。

与浏览器的 DOM 元素不同，**React 元素是创建开销极小的普通对象**。React DOM 会负责更新 DOM 来与 React 元素保持一致。

React 元素是不可变对象。一旦被创建，你就无法更改它的子元素或者属性。一个元素就像电影的单帧：它代表了某个特定时刻的 UI。

根据我们已有的知识，更新 UI 唯一的方式是创建一个全新的元素，并将其传入 ReactDOM.render()。

### 组件&props

所有 React 组件都必须像纯函数一样保护它们的 props 不被更改。

### State & 生命周期

#### 正确使用 State

+ 不要直接修改 State
+ State 的更新可能是异步的

出于性能考虑，React可能会把多个setState() 调用合并成一个调用。
因为 this.props 和 this.state 可能会异步更新，所以你不要依赖他们的值来更新下一个状态。

```javascript
// wrong
this.setState({
  counter: this.state.counter + this.props.increment.
})
```

要解决这个问题，可以让 setState() 接收一个函数而不是一个对象。这个函数用上一个 state 作为第一个参数，将此次更新被应用时的 props 做为第二个参数：

```javascript
// Correct
this.setState((state, props) => {
  counter: state.counter + props.increment,
})
```
+ State 的更新会被合并


### 事件处理

