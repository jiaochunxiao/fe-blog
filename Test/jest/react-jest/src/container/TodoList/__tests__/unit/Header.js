import React from 'react';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Header from '../../components/Header';

Enzyme.configure({
    adapter: new Adapter(),
});

it('Header渲染正常', () => {
    const wrapper = shallow(<Header />);
    expect(wrapper).toMatchSnapshot();
});

test('Header包含一个 input 框', () => {
    const wrapper = shallow(<Header />);
    const inputElem = wrapper.find('[data-test="input"]');
    expect(inputElem.length).toBe(1);
});

test('Header 组件input 框 值为空', () => {
    const wrapper = shallow(<Header />);
    const inputElem = wrapper.find('[data-test="input"]');
    expect(inputElem.prop('value')).toEqual('');
});

test('Header 组件input 框内容，用户输入时，会跟随变化', () => {
    const wrapper = shallow(<Header />);
    const inputElem = wrapper.find('[data-test="input"]');
    const userInput = '今天要学习jest';
    inputElem.simulate('change', {
        target: { value: userInput }
    });
    expect(wrapper.state('value')).toEqual(userInput);

    // const newInput = wrapper.find('[data-test="input"]');
    // expect(newInput.prop('value')).toBe(userInput);
});


test('Header 组件 input 框输入回车时，如果input 无内容，无操作', () => {
    const fn = jest.fn();
    const wrapper = shallow(<Header addUndoItem={fn} />);
    const inputElem = wrapper.find('[data-test="input"]');
    wrapper.setState({value: ''});
    inputElem.simulate('keyup', {
        keyCode: 13,
    });
    expect(fn).not.toHaveBeenCalled();
});


test('Header 组件 input 框输入回车时，如果input 有内容，函数应被调用', () => {
    const fn = jest.fn();
    const wrapper = shallow(<Header addUndoItem={fn} />);
    const inputElem = wrapper.find('[data-test="input"]');
    const userInput = '学习React';
    wrapper.setState({value: userInput});
    inputElem.simulate('keyUp', {
        keyCode: 13,
    });
    expect(fn).toHaveBeenCalled();
    expect(fn).toHaveBeenLastCalledWith(userInput);
    expect(inputElem.prop('value')).toBe('');
});


test('Header 组件 input 框输入回车时，如果input有值，最后被清除掉', () => {
    const fn = jest.fn();
    const wrapper = shallow(<Header addUndoItem={fn} />);
    const inputElem = wrapper.find('[data-test="input"]');
    const userInput = '学习React';
    wrapper.setState({value: userInput});
    inputElem.simulate('keyUp', {
        keyCode: 13,
    });
    const newInputElem = wrapper.find('[data-test="input"]');
    expect(newInputElem.prop('value')).toBe('');
});

