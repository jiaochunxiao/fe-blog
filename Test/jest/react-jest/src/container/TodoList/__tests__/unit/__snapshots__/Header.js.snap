// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header渲染正常 1`] = `
<div
  className="header"
>
  <div
    className="header-content"
  >
    TodoList
    <input
      className="header-input"
      data-test="input"
      onChange={[Function]}
      onKeyUp={[Function]}
      placeholder="todo"
      type="text"
      value=""
    />
  </div>
</div>
`;
